crates/wasm_interp/src/call_stack.rs

```diff
@@ -0,0 +1,210 @@
+Big size reduction, since most locals are i32. And we're loading that word from is_64 anyway.
+When pushing/popping frames, move the current frame offset using a similar calculation.
+Not clear if this would be better! Stack access pattern is pretty cache-friendly anyway.
+*/
+
+impl<'a> CallStack<'a> {
+    pub fn new(arena: &'a Bump) -> Self {
+        CallStack {
+            return_addrs: Vec::with_capacity_in(256, arena),
+            frame_offsets: Vec::with_capacity_in(256, arena),
+            locals_data: Vec::with_capacity_in(16 * 256, arena),
+            is_float: BitVec::with_capacity(256),
+            is_64: BitVec::with_capacity(256),
+        }
+    }
+
+    /// On entering a Wasm call, save the return address, and make space for locals
+    pub fn push_frame(&mut self, return_addr: u32, local_groups: &[(u32, ValueType)]) {
+        self.return_addrs.push(return_addr);
+        let frame_offset = self.is_64.len();
+        self.frame_offsets.push(frame_offset as u32);
+        let mut total = 0;
+        for (num_locals, ty) in local_groups {
+            let n = *num_locals as usize;
+            total += n;
+            self.is_64
+                .extend(repeat(matches!(ty, ValueType::I64 | ValueType::F64)).take(n));
+            self.is_float
+                .extend(repeat(matches!(ty, ValueType::F32 | ValueType::F64)).take(n));
+        }
+        self.locals_data.extend(repeat(0).take(total));
+    }
+
+    /// On returning from a Wasm call, drop its locals and retrieve the return address
+    pub fn pop_frame(&mut self) -> u32 {
+        let frame_offset = self.frame_offsets.pop().unwrap() as usize;
+        self.locals_data.truncate(frame_offset);
+        self.is_64.truncate(frame_offset);
+        self.is_64.truncate(frame_offset);
+        self.return_addrs.pop().unwrap()
+    }
+
+    pub fn get_local(&self, local_index: u32) -> Value {
+        let frame_offset = self.frame_offsets.last().unwrap();
+        let index = (*frame_offset + local_index) as usize;
+        let data64 = self.locals_data[index];
+        let is_float = self.is_float[index];
+        let is_64 = self.is_64[index];
+        if is_64 {
+            if is_float {
+                Value::F64(f64::from_ne_bytes(data64.to_ne_bytes()))
+            } else {
+                Value::I64(i64::from_ne_bytes(data64.to_ne_bytes()))
+            }
+        } else {
+            let data32 = data64 as u32;
+            if is_float {
+                Value::F32(f32::from_ne_bytes(data32.to_ne_bytes()))
+            } else {
+                Value::I32(i32::from_ne_bytes(data32.to_ne_bytes()))
+            }
+        }
+    }
+
+    pub fn set_local(&mut self, local_index: u32, value: Value) {
+        let frame_offset = *self.frame_offsets.last().unwrap();
+        let index = (frame_offset + local_index) as usize;
+        match value {
+            Value::I32(x) => {
+                self.locals_data[index] = unsafe { std::mem::transmute(x as i64) };
```

Comment:
I get it, but I think `u64::from_ne_bytes((x as i64).to_ne_bytes()) ` would be neater here (and less confusing)

Reply:
Yeah true. Now that you mention it, I did get confused by this myself at some point reading it back!
And there's no `unsafe` anywhere in the crate now.